### PR TITLE
Personal/sowu/version validation

### PR DESCRIPTION
**Pull request by sowu880**

### CODE REVIEW

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 1]
```diff
@@ -3,13 +3,34 @@
...
+            if (!string.IsNullOrEmpty(config.ConfigVersion))
+            {      
+                if (!string.Equals(allowedVersion[version], config.ConfigVersion))
+                {
+                    throw new AnonymizerConfigurationErrorsException($"The version of configuration is {config.ConfigVersion} where the {allowedVersion[version]} executable file is running");                  
+                }
+            }
+            else
+            {
+                _logger.LogWarning($"No version specification of configuration file, the version of {allowedVersion[version]} is assumed here. Mistakes may accured by this assumption");
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:31]: "Version is not specified in configuration..."

File: src/Fhir.Anonymizer.R4.Core.UnitTests/ConfigurationVersionTests.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    public class ConfigurationVersionTests
+    {
+        public ConfigurationVersionTests()
+        {
+            FhirPathCompiler.DefaultSymbolTable.AddExtensionSymbols();
+        }
+
+        public static IEnumerable<object[]> GetInvalidVersionOfConfigs()
+        {
+
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply src/Fhir.Anonymizer.R4.Core.UnitTests/ConfigurationVersionTests.cs:19]: The naming looks that configuration itself has multiple versions. We can name it "GetConfigsWithInvalidFhirVersion".

File: src/Fhir.Anonymizer.Stu3.Core.UnitTests/ConfigurationVersionTests.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            yield return new object[] { "./TestConfigurationsVersion/configuration-Stu3-version.json" };
+            yield return new object[] { "./TestConfigurationsVersion/configuration-empty-version.json" };
+            yield return new object[] { "./TestConfigurationsVersion/configuration-null-version.json" };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetInvalidVersionOfConfigs))]
+        public void GivenAnInvalidVersion_WhenCreateAnonymizerConfigurationManager_ExceptionShouldBeThrown(string configFilePath)
+        {
+            Assert.Throws<AnonymizerConfigurationErrorsException>(() => AnonymizerConfigurationManager.CreateFromConfigurationFile(configFilePath));
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Stu3.Core.UnitTests/ConfigurationVersionTests.cs:36]: We can use validator.validate() in unit tests directly, as the code change is in validator.

File: src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurations/configuration-test-sample.json [hunk 4]
```diff
@@ -19,8 +20,8 @@
     {
       "path": "Device.where(id.exists()).id",
       "method": "keep"
-    },
-    
+    }
+
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurations/configuration-test-sample.json:24]: extra empty lines in this and following configs

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 5]
```diff
@@ -3,13 +3,34 @@
...
+using Microsoft.Extensions.Logging;
+using Hl7.Fhir.Model;
+using System.Collections.Generic;
 
 namespace Fhir.Anonymizer.Core.AnonymizerConfigurations
 {
     public class AnonymizerConfigurationValidator
     {
+        private readonly ILogger _logger = AnonymizerLogging.CreateLogger<AnonymizerConfigurationValidator>();
+        private readonly Dictionary<string, string> allowedVersion= new Dictionary<string,string>{ { "3", "Stu3" },{ "4", "R4" } };
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:16]: Not sure if we have a better data structure to host the supported versions.

[the remark this answers is not in the document]
Comment (BoyaWu10) [reply src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:16]: Besides, we can move this into Constants.

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfiguration.cs [hunk 6]
```diff
@@ -7,6 +7,9 @@ namespace Fhir.Anonymizer.Core.AnonymizerConfigurations
     [DataContract]
     public class AnonymizerConfiguration
     {
+        [DataMember(Name = "version")]
+        public string ConfigVersion { get; set; }
```
[the remark this answers is not in the document]
Comment (moria97) [reply src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfiguration.cs:11]: +1

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 7]
```diff
@@ -3,13 +3,34 @@
...
         public void Validate(AnonymizerConfiguration config)
         {
+            var currentFullVersion = ModelInfo.Version;
+            string version = currentFullVersion.Split('.')[0];
```
[the remark this answers is not in the document]
Comment (moria97) [reply src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:17]: +1 We should handle the index out of range case here and check currentFullVersion is not null

File: src/Fhir.Anonymizer.Shared.Core/Constants.cs [hunk 8]
```diff
@@ -30,5 +28,7 @@ internal static class Constants
 
         internal const string GeneralResourceType = "Resource";
         internal const string GeneralDomainResourceType = "DomainResource";
+
+        internal static Dictionary<string, string> allowedVersion = new Dictionary<string, string> { { "3", "Stu3" }, { "4", "R4" } };
```
Comment (moria97) [inline src/Fhir.Anonymizer.Shared.Core/Constants.cs:32]: For now it is safe. But we should keep in mind that there will be a R5 update in the future

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 9]
```diff
@@ -1,15 +1,32 @@
...
     {
+        private readonly ILogger _logger = AnonymizerLogging.CreateLogger<AnonymizerConfigurationValidator>();
+        
         public void Validate(AnonymizerConfiguration config)
         {
+            var currentFullVersion = ModelInfo.Version;
+            string version = currentFullVersion.Split('.')[0];
+            if (!string.IsNullOrEmpty(config.ConfigVersion))
+            {      
+                if (!string.Equals(Constants.allowedVersion[version], config.ConfigVersion))
```
Comment (moria97) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:20]: We need handle the corner case where version is not in allowedVersion

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 10]
```diff
@@ -1,15 +1,32 @@
...
+                {
+                    throw new AnonymizerConfigurationErrorsException($"The version of executable and configuration do not match: executable version is {Constants.allowedVersion[version]} but configuration version is {config.ConfigVersion}.");
```
Comment (moria97) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:22]: To be concise here: "Configuration of fhirVersion {config.FhirVersion} not supported. Expected fhirVersion: {constants.allowedVersion[version]}".

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 11]
```diff
@@ -1,15 +1,32 @@
...
+            if (!string.IsNullOrEmpty(config.ConfigVersion))
+            {      
+                if (!string.Equals(Constants.allowedVersion[version], config.ConfigVersion))
+                {
+                    throw new AnonymizerConfigurationErrorsException($"The version of executable and configuration do not match: executable version is {Constants.allowedVersion[version]} but configuration version is {config.ConfigVersion}.");                  
+                }
+            }
+            else
+            {
+                _logger.LogWarning($"Version is not specified in configuration file, the version of {Constants.allowedVersion[version]} is assumed here. Mistakes may accured by this assumption");
```
Comment (moria97) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:27]: Error message would be better to be concise.

File: src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurationsVersion/configuration-Stu3-version.json [hunk 12]
```diff
@@ -0,0 +1,49 @@
+﻿{
+  "fhirVersion": "Stu3",
```
Comment (Patrick2011) [inline src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurationsVersion/configuration-Stu3-version.json:2]: Is this case sensitive?

Comment (sowu880) [reply src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurationsVersion/configuration-Stu3-version.json:2]: yes. Is it necessary?

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 13]
```diff
@@ -1,15 +1,44 @@
...
+                    }   
+                }
+                else
+                {
+                    throw new AnonymizerConfigurationErrorsException("Fail to read supported FHIR version");
+                }    
+            }
+            else
+            {
+                _logger.LogWarning($"Version is not specified in configuration file. Mistakes may accured by missing version");
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:39]: We can remove the second sentence.

File: src/Fhir.Anonymizer.Shared.Core/Constants.cs [hunk 14]
```diff
@@ -30,5 +28,7 @@ internal static class Constants
 
         internal const string GeneralResourceType = "Resource";
         internal const string GeneralDomainResourceType = "DomainResource";
+
+        internal enum allowedVersion { Stu3=3, R4=4 };
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Shared.Core/Constants.cs:32]: split it into multiple lines

Comment (BoyaWu10) [reply src/Fhir.Anonymizer.Shared.Core/Constants.cs:32]: and we can name it supportedVersion

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 15]
```diff
@@ -1,15 +1,44 @@
...
+            if (!string.IsNullOrEmpty(config.FhirVersion))
+            {
+                var currentFullVersion = ModelInfo.Version;
+                if (string.IsNullOrEmpty(currentFullVersion))
+                {
+                    throw new AnonymizerConfigurationErrorsException("Fail to read supported FHIR version");
+                }
+                string versionNumber = currentFullVersion.Split('.')[0];
+                Constants.allowedVersion versionCode;
+                if (Enum.TryParse(versionNumber, true, out versionCode) && Enum.IsDefined(typeof(Constants.allowedVersion), versionCode))
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:25]: Is Enum.TryParse enough to parse the variable?

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 16]
```diff
@@ -1,15 +1,44 @@
...
+                {
+                    if (!string.Equals(versionCode.ToString(), config.FhirVersion))
+                    {
+                        throw new AnonymizerConfigurationErrorsException($"Configuration of fhirVersion {config.FhirVersion} is not supported. Expected fhirVersion: {versionCode}");
+                    }   
+                }
+                else
+                {
+                    throw new AnonymizerConfigurationErrorsException("Fail to read supported FHIR version");
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:34]: Could we share the same error message as Line 29?

File: src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurationsVersion/configuration-Stu3-version.json (hunk 12 above)
Comment (sowu880) [reply src/Fhir.Anonymizer.Shared.Core.UnitTest/TestConfigurationsVersion/configuration-Stu3-version.json:2]: Changed it to case insensitive.

File: README.md [hunk 17]
```diff
@@ -50,9 +50,16 @@ This repo contains a few [sample](samples/fhir-r4-files) FHIR files that you can
 You can also export FHIR resource from your FHIR server using [Bulk Export](https://docs.microsoft.com/en-us/azure/healthcare-apis/configure-export-data).
 
 ## Anonymize FHIR data using the command line tool
-Once you have built the command line tool, you will find the Fhir.Anonymizer.Tool.exe in the $SOURCE\src\Fhir.Anonymizer.Tool\bin\Debug|Release\netcoreapp3.1 folder. You can use this exe to anonymize FHIR resource files in a folder.
```
Comment (moria97) [inline README.md:53]: Can we remove the doc change in this PR?

File: src/Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json [hunk 18]
```diff
@@ -1,200 +1,771 @@
 {
+  "fhirVersion": "R4",
   "fhirPathRules": [
-    {"path": "nodesByType('Extension')", "method": "redact"},
```
Comment (moria97) [inline src/Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json:3]: While adding "R4", we should make the rules in one line to shrink the config size

File: src/Fhir.Anonymizer.R4.Core/Constants.cs [hunk 19]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace Fhir.Anonymizer.Core
+{
+    internal static partial class Constants
+    {
+        // InstanceType constants
+        internal const string supportedVersion = "R4";
```
Comment (moria97) [inline src/Fhir.Anonymizer.R4.Core/Constants.cs:8]: The recommended naming and capitalization convention is to use PascalCasing for constants

File: src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs [hunk 20]
```diff
@@ -1,15 +1,27 @@
...
         public void Validate(AnonymizerConfiguration config)
-        {
+        {   
+            if (!string.IsNullOrEmpty(config.FhirVersion) && !string.Equals(Constants.supportedVersion, config.FhirVersion, StringComparison.InvariantCultureIgnoreCase))
+            {
+                throw new AnonymizerConfigurationErrorsException($"Configuration of fhirVersion {config.FhirVersion} is not supported. Expected fhirVersion: {Constants.supportedVersion}");
+            }
+            else
+            {
+                _logger.LogWarning($"Version is not specified in configuration file.");
```
Comment (moria97) [inline src/Fhir.Anonymizer.Shared.Core/AnonymizationConfigurations/AnonymizerConfigurationValidator.cs:22]: here seems would be a warning if the FhirVersion is equal to Constants.SupportedVersion

**Pull request merged**